File: ciel/evolution_hydra/EntropyHarvester.ts
import logger from '../utils/logger.js';
import { HardwareMetrics } from '../utils/HardwareMetrics.js';
import type { Genome } from './Genome.js';

/**
 * ═══════════════════════════════════════════════════════════════
 * ENTROPY HARVESTER — Creative Disorder Engine
 * ═══════════════════════════════════════════════════════════════
 *
 * PRINCIPE RÉVOLUTIONNAIRE :
 * L'entropie est l'ennemi de la plupart des systèmes — mais pas
 * de HYDRA. L'EntropyHarvester CAPTURE le désordre naturel du
 * système (CPU fluctuations, erreurs, timing variances) et le
 * TRANSFORME en créativité computationnelle.
 *
 * Fondements théoriques :
 * - Thermodynamique hors équilibre (Prigogine) : les structures
 *   dissipatives UTILISENT le désordre pour s'auto-organiser
 * - Bruit stochastique comme ressource (Stochastic Resonance) :
 *   un certain niveau de bruit AMÉLIORE la détection du signal
 * - Antifragilité (Taleb) : les systèmes qui profitent du désordre
 * - Processus de Poisson comme source de nouveauté
 * - Mutation dirigée par l'entropie : au lieu de combattre
 *   le bruit, l'utiliser comme moteur d'exploration
 *
 * Le Harvester ne produit pas du chaos — il DISTILLE le chaos
 * naturel en VARIATION CRÉATIVE. C'est un alchimiste computationnel
 * qui transforme le plomb du bruit en or de la nouveauté.
 */

// ─── Types de l'Harvester ───────────────────────────────────

export interface EntropySource {
  id: string;
  type: EntropySourceType;
  rawValue: number;
  normalizedValue: number;       // 0-1
  timestamp: number;
  quality: number;               // Qualité de l'entropie (0-1)
}

export enum EntropySourceType {
  CPU_FLUCTUATION = 'CPU_FLUCTUATION',
  MEMORY_VARIANCE = 'MEMORY_VARIANCE',
  EVENT_LOOP_LAG = 'EVENT_LOOP_LAG',
  TIMING_JITTER = 'TIMING_JITTER',
  MUTATION_DRIFT = 'MUTATION_DRIFT',
  NETWORK_LATENCY = 'NETWORK_LATENCY',
  DISK_IO_VARIANCE = 'DISK_IO_VARIANCE',
  THERMAL_NOISE = 'THERMAL_NOISE',
}

export interface HarvestedEntropy {
  id: string;
  sources: EntropySource[];
  compositeValue: number;        // Valeur composite normalisée
  creativePotential: number;     // Potentiel créatif (0-1)
  entropyType: EntropyProfile;
  harvestedAt: number;
  usedFor: string | null;        // Comment cette entropie a été utilisée
}

export enum EntropyProfile {
  CHAOTIC = 'CHAOTIC',           // Entropie pure, non structurée
  STRUCTURED = 'STRUCTURED',     // Entropie avec des patterns sous-jacents
  RESONANT = 'RESONANT',         // Entropie en résonance avec le système
  CATALYTIC = 'CATALYTIC',       // Entropie qui accélère l'évolution
  DORMANT = 'DORMANT',           // Entropie stockée pour usage futur
}

export interface EntropyState {
  totalHarvested: number;
  totalUsed: number;
  harvestRate: number;           // Entropie récoltée par seconde
  creativeYield: number;         // Rendement créatif (0-1)
  dominantProfile: EntropyProfile;
  stochasticResonanceLevel: number; // Niveau de résonance stochastique
  antifragilityIndex: number;    // Indice d'antifragilité du système
}

// ─── Le Récolteur d'Entropie ────────────────────────────────

export class EntropyHarvester {
  private sources: Map<EntropySourceType, EntropySource> = new Map();
  private harvested: HarvestedEntropy[] = [];
  private entropyPool: number[] = [];  // Pool d'entropie brute
  private state: EntropyState;
  private lastHarvestTime: number = Date.now();
  private stochasticResonanceOptimum: number = 0.3; // Niveau optimal de bruit

  constructor() {
    this.state = {
      totalHarvested: 0,
      totalUsed: 0,
      harvestRate: 0,
      creativeYield: 0,
      dominantProfile: EntropyProfile.DORMANT,
      stochasticResonanceLevel: 0,
      antifragilityIndex: 0.5,
    };

    logger.info('[Entropy Harvester] 🌊 Creative Disorder Engine initialized. Chaos is not the enemy — it is the fuel.');
  }

  // ─── Récolte de l'Entropie ─────────────────────────────

  /**
   * Récolte l'entropie naturelle du système.
   * Chaque source de variation est capturée, normalisée, et
   * évaluée pour sa qualité créative.
   */
  harvest(): HarvestedEntropy {
    const now = Date.now();

    // 1. Capturer les sources d'entropie
    const sources: EntropySource[] = [];

    // CPU Fluctuation
    const cpuLoad = HardwareMetrics.getRealCPULoad();
    sources.push(this.createSource(EntropySourceType.CPU_FLUCTUATION, cpuLoad));

    // Memory Variance
    const memUsage = HardwareMetrics.getRealMemoryUsage();
    sources.push(this.createSource(EntropySourceType.MEMORY_VARIANCE, memUsage));

    // Event Loop Lag
    const lag = HardwareMetrics.getEventLoopLag();
    sources.push(this.createSource(EntropySourceType.EVENT_LOOP_LAG, lag / 100)); // Normaliser

    // Timing Jitter
    const jitter = Math.abs((now - this.lastHarvestTime) - 1000) / 1000; // Variation par rapport à 1s
    sources.push(this.createSource(EntropySourceType.TIMING_JITTER, Math.min(1, jitter)));

    // Thermal Noise (simulation basée sur les fluctuations CPU)
    const thermalNoise = Math.abs(cpuLoad - 0.5) * 2; // Plus c'est loin de 0.5, plus il y a de chaleur
    sources.push(this.createSource(EntropySourceType.THERMAL_NOISE, thermalNoise));

    this.lastHarvestTime = now;

    // 2. Calculer la valeur composite
    const compositeValue = this.computeCompositeEntropy(sources);

    // 3. Déterminer le profil d'entropie
    const entropyType = this.classifyEntropy(sources, compositeValue);

    // 4. Calculer le potentiel créatif
    const creativePotential = this.computeCreativePotential(sources, entropyType);

    // 5. Créer l'entrée récoltée
    const harvested: HarvestedEntropy = {
      id: `entropy_${Date.now()}_${Math.random().toString(36).slice(2, 5)}`,
      sources,
      compositeValue,
      creativePotential,
      entropyType,
      harvestedAt: now,
      usedFor: null,
    };

    this.harvested.push(harvested);
    this.entropyPool.push(compositeValue);
    this.state.totalHarvested++;

    // Garder le pool à une taille raisonnable
    if (this.entropyPool.length > 1000) {
      this.entropyPool = this.entropyPool.slice(-500);
    }

    // Mettre à jour les stats
    this.updateStats(harvested);

    logger.debug(`[Entropy Harvester] 🌊 Harvested entropy: ${compositeValue.toFixed(3)} (type: ${entropyType}, creative: ${creativePotential.toFixed(3)})`);

    return harvested;
  }

  // ─── Utilisation de l'Entropie ─────────────────────────

  /**
   * Utilise l'entropie récoltée pour alimenter un processus créatif.
   * L'entropie est dirigée vers là où elle sera la plus productive.
   */
  useEntropy(purpose: string, requiredAmount: number = 0.3): HarvestedEntropy | null {
    // Trouver l'entrée récoltée la plus appropriée
    let bestMatch: HarvestedEntropy | null = null;
    let bestScore = 0;

    for (const h of this.harvested) {
      if (h.usedFor !== null) continue; // Déjà utilisée

      // Score basé sur la proximité du potentiel créatif avec le besoin
      const match = 1 - Math.abs(h.creativePotential - requiredAmount);

      // Bonus pour les profils catalytiques et résonants
      const profileBonus = h.entropyType === EntropyProfile.CATALYTIC ? 0.3 :
                           h.entropyType === EntropyProfile.RESONANT ? 0.2 : 0;

      const score = match + profileBonus;

      if (score > bestScore) {
        bestScore = score;
        bestMatch = h;
      }
    }

    if (bestMatch) {
      bestMatch.usedFor = purpose;
      this.state.totalUsed++;

      logger.debug(`[Entropy Harvester] ⚡ Entropy USED for: ${purpose} (creative: ${bestMatch.creativePotential.toFixed(3)})`);
    }

    return bestMatch;
  }

  /**
   * Génère une mutation créative basée sur l'entropie récoltée.
   * Au lieu de mutations aléatoires pures, les mutations sont
   * GUIDÉES par l'entropie structurée.
   */
  generateCreativeMutation(genome: Genome): {
    targetParam: string;
    delta: number;
    rationale: string;
  } | null {
    const entropy = this.useEntropy(`mutation_${genome.id}`, 0.5);
    if (!entropy) return null;

    // Choisir le paramètre à muter basé sur le profil d'entropie
    let targetParam: string;
    let delta: number;
    let rationale: string;

    switch (entropy.entropyType) {
      case EntropyProfile.CHAOTIC:
        // L'entropie chaotique pousse vers des mutations exploratoires
        targetParam = 'temperature';
        delta = entropy.compositeValue * 0.5 - 0.25; // ±0.25
        rationale = `Chaotic entropy surge driving temperature exploration (Δ=${delta.toFixed(3)})`;
        break;

      case EntropyProfile.STRUCTURED:
        // L'entropie structurée guide vers des mutations de spécialisation
        targetParam = 'promptMutation';
        const traits = [
          'Be highly analytical and concise.',
          'Think step-by-step and write pseudocode before coding.',
          'Propose radically different alternative solutions.',
        ];
        delta = 0; // Pour promptMutation, delta n'est pas utilisé
        rationale = traits[Math.floor(entropy.compositeValue * traits.length)];
        targetParam = 'promptMutation';
        break;

      case EntropyProfile.RESONANT:
        // L'entropie résonante amplifie les forces existantes
        const tools = Object.keys(genome.params.toolWeights);
        targetParam = `tool_${tools[Math.floor(entropy.compositeValue * tools.length)]}`;
        delta = entropy.creativePotential * 0.3; // Amplification positive
        rationale = `Resonant entropy amplifying existing tool affinity (${targetParam})`;
        break;

      case EntropyProfile.CATALYTIC:
        // L'entropie catalytique provoque des sauts qualitatifs
        targetParam = 'polyglotDepth';
        delta = entropy.compositeValue * 0.4; // Saut significatif
        rationale = `Catalytic entropy triggering qualitative leap in polyglot capability`;
        break;

      default:
        targetParam = 'episodicRetention';
        delta = entropy.compositeValue * 0.2 - 0.1;
        rationale = `Dormant entropy gently adjusting memory retention`;
    }

    return { targetParam, delta, rationale };
  }

  // ─── Résonance Stochastique ────────────────────────────

  /**
   * La résonance stochastique se produit quand un niveau OPTIMAL
   * de bruit AMÉLIORE la performance du système.
   * L'Harvester surveille ce niveau et ajuste l'apport en bruit.
   */
  computeStochasticResonance(systemPerformance: number): number {
    // La performance du système en fonction du bruit suit une courbe en U inversé
    // Il y a un optimum de bruit où la performance est maximale
    const noiseLevel = this.state.harvestRate;
    const optimum = this.stochasticResonanceOptimum;

    // La résonance est maximale quand le bruit est proche de l'optimum
    const deviation = Math.abs(noiseLevel - optimum) / optimum;
    const resonance = Math.exp(-deviation * deviation * 2); // Courbe gaussienne

    this.state.stochasticResonanceLevel = resonance;

    // Ajuster l'optimum basé sur la performance observée
    if (systemPerformance > 0.7 && resonance > 0.5) {
      // Bonne performance avec bonne résonance → l'optimum est correct
    } else if (systemPerformance < 0.3 && noiseLevel > optimum) {
      // Mauvaise performance avec trop de bruit → réduire l'optimum
      this.stochasticResonanceOptimum *= 0.95;
    } else if (systemPerformance < 0.3 && noiseLevel < optimum) {
      // Mauvaise performance avec trop peu de bruit → augmenter l'optimum
      this.stochasticResonanceOptimum = Math.min(0.8, this.stochasticResonanceOptimum * 1.05);
    }

    return resonance;
  }

  // ─── Antifragilité ─────────────────────────────────────

  /**
   * Mesure l'antifragilité du système : un système antifragile
   * PROFITE du désordre (il s'améliore quand il est stressé).
   */
  computeAntifragilityIndex(recentPerformanceHistory: number[]): number {
    if (recentPerformanceHistory.length < 5) return 0.5;

    // Comparer la performance après les périodes de haute entropie
    // vs après les périodes de basse entropie
    const half = Math.floor(recentPerformanceHistory.length / 2);
    const firstHalf = recentPerformanceHistory.slice(0, half);
    const secondHalf = recentPerformanceHistory.slice(half);

    const avgFirst = firstHalf.reduce((a, b) => a + b, 0) / firstHalf.length;
    const avgSecond = secondHalf.reduce((a, b) => a + b, 0) / secondHalf.length;

    // Si le système s'améliore avec le temps (même avec le stress), il est antifragile
    const improvement = avgSecond - avgFirst;

    // Si le système se dégrade avec le stress, il est fragile
    this.state.antifragilityIndex = Math.tanh(improvement * 5 + 0.5);

    if (this.state.antifragilityIndex > 0.7) {
      logger.info(`[Entropy Harvester] 💪 ANTIFRAGILE: System improves under stress (index: ${this.state.antifragilityIndex.toFixed(3)})`);
    } else if (this.state.antifragilityIndex < 0.3) {
      logger.warn(`[Entropy Harvester] ⚠️ FRAGILE: System degrades under stress (index: ${this.state.antifragilityIndex.toFixed(3)})`);
    }

    return this.state.antifragilityIndex;
  }

  // ─── Utilitaires ────────────────────────────────────────

  private createSource(type: EntropySourceType, rawValue: number): EntropySource {
    return {
      id: `src_${type}_${Date.now()}`,
      type,
      rawValue,
      normalizedValue: Math.max(0, Math.min(1, rawValue)),
      timestamp: Date.now(),
      quality: this.assessSourceQuality(type, rawValue),
    };
  }

  private assessSourceQuality(type: EntropySourceType, value: number): number {
    // La qualité dépend de la source et de la valeur
    // Les valeurs extrêmes (proches de 0 ou 1) ont souvent une meilleure qualité
    // car elles sont moins prévisibles
    switch (type) {
      case EntropySourceType.CPU_FLUCTUATION:
        return value > 0.3 && value < 0.9 ? 0.8 : 0.4; // Zone productive
      case EntropySourceType.THERMAL_NOISE:
        return value > 0.5 ? 0.9 : 0.3; // Plus de chaleur = plus d'entropie
      case EntropySourceType.TIMING_JITTER:
        return value > 0.1 ? 0.7 : 0.2; // Le jitter est une bonne source
      default:
        return 0.5;
    }
  }

  private computeCompositeEntropy(sources: EntropySource[]): number {
    // Moyenne pondérée par la qualité
    let totalWeight = 0;
    let weightedSum = 0;

    for (const source of sources) {
      weightedSum += source.normalizedValue * source.quality;
      totalWeight += source.quality;
    }

    return totalWeight > 0 ? weightedSum / totalWeight : 0;
  }

  private classifyEntropy(sources: EntropySource[], composite: number): EntropyProfile {
    // Variance entre les sources
    const values = sources.map(s => s.normalizedValue);
    const avg = values.reduce((a, b) => a + b, 0) / values.length;
    const variance = values.reduce((acc, v) => acc + Math.pow(v - avg, 2), 0) / values.length;

    // Haute variance = chaotique
    // Basse variance = structuré
    // Variance modérée avec haute qualité = résonant
    // Composite élevé = catalytique

    const avgQuality = sources.reduce((acc, s) => acc + s.quality, 0) / sources.length;

    if (composite > 0.7 && avgQuality > 0.6) return EntropyProfile.CATALYTIC;
    if (variance < 0.05 && avgQuality > 0.5) return EntropyProfile.RESONANT;
    if (variance > 0.15) return EntropyProfile.CHAOTIC;
    if (variance < 0.05) return EntropyProfile.STRUCTURED;
    return EntropyProfile.DORMANT;
  }

  private computeCreativePotential(sources: EntropySource[], profile: EntropyProfile): number {
    // Le potentiel créatif dépend du profil et de la qualité
    const profileMultiplier: Record<EntropyProfile, number> = {
      [EntropyProfile.CATALYTIC]: 1.0,
      [EntropyProfile.RESONANT]: 0.8,
      [EntropyProfile.CHAOTIC]: 0.6,
      [EntropyProfile.STRUCTURED]: 0.5,
      [EntropyProfile.DORMANT]: 0.2,
    };

    const avgQuality = sources.reduce((acc, s) => acc + s.quality, 0) / sources.length;
    const composite = this.computeCompositeEntropy(sources);

    return Math.min(1.0, composite * avgQuality * (profileMultiplier[profile] || 0.5));
  }

  private updateStats(latestHarvest: HarvestedEntropy): void {
    // Taux de récolte
    const timeSinceLastHarvest = (Date.now() - this.lastHarvestTime) / 1000;
    this.state.harvestRate = timeSinceLastHarvest > 0
      ? 1 / timeSinceLastHarvest
      : 0;

    // Rendement créatif
    const usedEntries = this.harvested.filter(h => h.usedFor !== null);
    this.state.creativeYield = this.state.totalHarvested > 0
      ? usedEntries.length / this.state.totalHarvested
      : 0;

    // Profil dominant
    const profileCounts: Record<EntropyProfile, number> = {
      [EntropyProfile.CHAOTIC]: 0,
      [EntropyProfile.STRUCTURED]: 0,
      [EntropyProfile.RESONANT]: 0,
      [EntropyProfile.CATALYTIC]: 0,
      [EntropyProfile.DORMANT]: 0,
    };

    for (const h of this.harvested.slice(-100)) {
      profileCounts[h.entropyType]++;
    }

    let maxCount = 0;
    for (const [profile, count] of Object.entries(profileCounts)) {
      if (count > maxCount) {
        maxCount = count;
        this.state.dominantProfile = profile as EntropyProfile;
      }
    }
  }

  // ─── Getters ────────────────────────────────────────────

  getState(): EntropyState {
    return { ...this.state };
  }

  getEntropyPool(): number[] {
    return [...this.entropyPool];
  }

  getRecentHarvests(limit: number = 50): HarvestedEntropy[] {
    return this.harvested.slice(-limit);
  }
}

File: ciel/evolution/entropy_harvester.py
"""
EntropyHarvester - Creative Disorder Engine
Captures system entropy and transforms it into creative potential.
"""
from __future__ import annotations

import logging
import math
import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Optional

logger = logging.getLogger(__name__)


class EntropySourceType(str, Enum):
    """Types of entropy sources"""
    CPU_FLUCTUATION = "CPU_FLUCTUATION"
    MEMORY_VARIANCE = "MEMORY_VARIANCE"
    EVENT_LOOP_LAG = "EVENT_LOOP_LAG"
    TIMING_JITTER = "TIMING_JITTER"
    MUTATION_DRIFT = "MUTATION_DRIFT"
    NETWORK_LATENCY = "NETWORK_LATENCY"
    DISK_IO_VARIANCE = "DISK_IO_VARIANCE"
    THERMAL_NOISE = "THERMAL_NOISE"


class EntropyProfile(str, Enum):
    """Entropy classifications"""
    CHAOTIC = "CHAOTIC"
    STRUCTURED = "STRUCTURED"
    RESONANT = "RESONANT"
    CATALYTIC = "CATALYTIC"
    DORMANT = "DORMANT"


@dataclass(slots=True)
class EntropySource:
    """Individual entropy source"""
    id: str
    type: EntropySourceType
    raw_value: float
    normalized_value: float  # 0-1
    timestamp: float
    quality: float  # 0-1


@dataclass(slots=True)
class HarvestedEntropy:
    """Harvested entropy record"""
    id: str
    sources: list[EntropySource]
    composite_value: float
    creative_potential: float  # 0-1
    entropy_type: EntropyProfile
    harvested_at: float
    used_for: Optional[str] = None


@dataclass(slots=True)
class EntropyState:
    """State of entropy harvester"""
    total_harvested: int = 0
    total_used: int = 0
    harvest_rate: float = 0.0
    creative_yield: float = 0.0
    dominant_profile: EntropyProfile = EntropyProfile.DORMANT
    stochastic_resonance_level: float = 0.0
    antifragility_index: float = 0.5


@dataclass(slots=True)
class EntropyHarvester:
    """
    EntropyHarvester - Creative Disorder Engine
    Transforms system entropy into creative potential.
    """
    sources: dict[EntropySourceType, EntropySource] = field(default_factory=dict)
    harvested: list[HarvestedEntropy] = field(default_factory=list)
    entropy_pool: list[float] = field(default_factory=list)
    state: EntropyState = field(default_factory=EntropyState)
    last_harvest_time: float = field(default_factory=time.time)
    stochastic_resonance_optimum: float = 0.3

    def __post_init__(self) -> None:
        """Initialize entropy harvester"""
        logger.info(
            "[Entropy Harvester] 🌊 Creative Disorder Engine initialized. "
            "Chaos is not the enemy — it is the fuel."
        )

    def harvest(self) -> HarvestedEntropy:
        """Harvest entropy from system"""
        now = time.time()

        sources: list[EntropySource] = []

        # Simulate CPU fluctuation
        cpu_load = (hash(str(now)) % 100) / 100
        sources.append(self._create_source(EntropySourceType.CPU_FLUCTUATION, cpu_load))

        # Memory variance
        mem_usage = (hash(str(now * 2)) % 100) / 100
        sources.append(self._create_source(EntropySourceType.MEMORY_VARIANCE, mem_usage))

        # Event loop lag
        lag = (hash(str(now * 3)) % 50) / 50
        sources.append(self._create_source(EntropySourceType.EVENT_LOOP_LAG, lag))

        # Timing jitter
        jitter = min(1, abs((now - self.last_harvest_time) - 1.0) / 1.0)
        sources.append(self._create_source(EntropySourceType.TIMING_JITTER, jitter))

        # Thermal noise
        thermal = abs(cpu_load - 0.5) * 2
        sources.append(self._create_source(EntropySourceType.THERMAL_NOISE, thermal))

        self.last_harvest_time = now

        composite_value = self._compute_composite_entropy(sources)
        entropy_type = self._classify_entropy(sources, composite_value)
        creative_potential = self._compute_creative_potential(sources, entropy_type)

        harvested = HarvestedEntropy(
            id=f"entropy_{int(now)}",
            sources=sources,
            composite_value=composite_value,
            creative_potential=creative_potential,
            entropy_type=entropy_type,
            harvested_at=now,
            used_for=None,
        )

        self.harvested.append(harvested)
        self.entropy_pool.append(composite_value)
        self.state.total_harvested += 1

        # Keep pool size reasonable
        if len(self.entropy_pool) > 1000:
            self.entropy_pool = self.entropy_pool[-500:]

        self._update_stats(harvested)

        logger.debug(
            f"[Entropy Harvester] 🌊 Harvested entropy: {composite_value:.3f} "
            f"(type: {entropy_type.value}, creative: {creative_potential:.3f})"
        )

        return harvested

    def use_entropy(self, purpose: str, required_amount: float = 0.3) -> Optional[HarvestedEntropy]:
        """Use harvested entropy for creative process"""
        best_match = None
        best_score = 0.0

        for h in self.harvested:
            if h.used_for is not None:
                continue

            match = 1 - abs(h.creative_potential - required_amount)
            profile_bonus = (
                0.3 if h.entropy_type == EntropyProfile.CATALYTIC else
                0.2 if h.entropy_type == EntropyProfile.RESONANT else
                0.0
            )

            score = match + profile_bonus

            if score > best_score:
                best_score = score
                best_match = h

        if best_match:
            best_match.used_for = purpose
            self.state.total_used += 1
            logger.debug(
                f"[Entropy Harvester] ⚡ Entropy USED for: {purpose} "
                f"(creative: {best_match.creative_potential:.3f})"
            )

        return best_match

    def compute_stochastic_resonance(self, system_performance: float) -> float:
        """Compute optimal noise level for system performance"""
        noise_level = self.state.harvest_rate
        optimum = self.stochastic_resonance_optimum

        deviation = abs(noise_level - optimum) / optimum
        resonance = math.exp(-deviation * deviation * 2)

        self.state.stochastic_resonance_level = resonance

        if system_performance > 0.7 and resonance > 0.5:
            pass  # Good performance, keep optimum
        elif system_performance < 0.3 and noise_level > optimum:
            self.stochastic_resonance_optimum *= 0.95
        elif system_performance < 0.3 and noise_level < optimum:
            self.stochastic_resonance_optimum = min(0.8, self.stochastic_resonance_optimum * 1.05)

        return resonance

    def compute_antifragility_index(self, recent_performance: list[float]) -> float:
        """Measure if system improves under stress"""
        if len(recent_performance) < 5:
            return 0.5

        half = len(recent_performance) // 2
        first_half = recent_performance[:half]
        second_half = recent_performance[half:]

        avg_first = sum(first_half) / len(first_half)
        avg_second = sum(second_half) / len(second_half)

        improvement = avg_second - avg_first
        self.state.antifragility_index = math.tanh(improvement * 5 + 0.5)

        if self.state.antifragility_index > 0.7:
            logger.info(
                f"[Entropy Harvester] 💪 ANTIFRAGILE: System improves under stress "
                f"(index: {self.state.antifragility_index:.3f})"
            )
        elif self.state.antifragility_index < 0.3:
            logger.warning(
                f"[Entropy Harvester] ⚠️ FRAGILE: System degrades under stress "
                f"(index: {self.state.antifragility_index:.3f})"
            )

        return self.state.antifragility_index

    def _create_source(
        self, source_type: EntropySourceType, raw_value: float
    ) -> EntropySource:
        """Create entropy source"""
        return EntropySource(
            id=f"src_{source_type.value}_{int(time.time())}",
            type=source_type,
            raw_value=raw_value,
            normalized_value=max(0, min(1, raw_value)),
            timestamp=time.time(),
            quality=self._assess_source_quality(source_type, raw_value),
        )

    def _assess_source_quality(
        self, source_type: EntropySourceType, value: float
    ) -> float:
        """Assess quality of entropy source"""
        match source_type:
            case EntropySourceType.CPU_FLUCTUATION:
                return 0.8 if 0.3 < value < 0.9 else 0.4
            case EntropySourceType.THERMAL_NOISE:
                return 0.9 if value > 0.5 else 0.3
            case EntropySourceType.TIMING_JITTER:
                return 0.7 if value > 0.1 else 0.2
            case _:
                return 0.5

    def _compute_composite_entropy(self, sources: list[EntropySource]) -> float:
        """Compute composite entropy value"""
        total_weight = sum(s.quality for s in sources)
        if total_weight == 0:
            return 0.0
        return sum(s.normalized_value * s.quality for s in sources) / total_weight

    def _classify_entropy(
        self, sources: list[EntropySource], composite: float
    ) -> EntropyProfile:
        """Classify entropy profile"""
        if not sources:
            return EntropyProfile.DORMANT

        values = [s.normalized_value for s in sources]
        avg = sum(values) / len(values)
        variance = sum((v - avg) ** 2 for v in values) / len(values)
        avg_quality = sum(s.quality for s in sources) / len(sources)

        if composite > 0.7 and avg_quality > 0.6:
            return EntropyProfile.CATALYTIC
        if variance < 0.05 and avg_quality > 0.5:
            return EntropyProfile.RESONANT
        if variance > 0.15:
            return EntropyProfile.CHAOTIC
        if variance < 0.05:
            return EntropyProfile.STRUCTURED
        return EntropyProfile.DORMANT

    def _compute_creative_potential(
        self, sources: list[EntropySource], profile: EntropyProfile
    ) -> float:
        """Compute creative potential"""
        multiplier = {
            EntropyProfile.CATALYTIC: 1.0,
            EntropyProfile.RESONANT: 0.8,
            EntropyProfile.CHAOTIC: 0.6,
            EntropyProfile.STRUCTURED: 0.5,
            EntropyProfile.DORMANT: 0.2,
        }.get(profile, 0.5)

        avg_quality = sum(s.quality for s in sources) / len(sources) if sources else 0
        composite = self._compute_composite_entropy(sources)

        return min(1.0, composite * avg_quality * multiplier)

    def _update_stats(self, latest: HarvestedEntropy) -> None:
        """Update statistics"""
        time_since = time.time() - self.last_harvest_time
        self.state.harvest_rate = 1 / time_since if time_since > 0 else 0

        used_entries = [h for h in self.harvested if h.used_for is not None]
        self.state.creative_yield = (
            len(used_entries) / self.state.total_harvested
            if self.state.total_harvested > 0 else 0
        )

        profile_counts = {profile: 0 for profile in EntropyProfile}
        for h in self.harvested[-100:]:
            profile_counts[h.entropy_type] += 1

        max_count = max(profile_counts.values()) if profile_counts else 0
        for profile, count in profile_counts.items():
            if count == max_count:
                self.state.dominant_profile = profile
                break

    def get_state(self) -> EntropyState:
        """Get current state"""
        return self.state

    def get_entropy_pool(self) -> list[float]:
        """Get entropy pool copy"""
        return self.entropy_pool.copy()

    def get_recent_harvests(self, limit: int = 50) -> list[HarvestedEntropy]:
        """Get recent harvests"""
        return self.harvested[-limit:]

    def process(self, input_data: Any) -> dict[str, Any]:
        """
        Process entropy request.
        CIEL compatibility method.
        """
        harvested = self.harvest()
        return {
            "harvested": harvested.composite_value,
            "creative_potential": harvested.creative_potential,
            "profile": harvested.entropy_type.value,
            "state": {
                "total_harvested": self.state.total_harvested,
                "antifragility": self.state.antifragility_index,
            }
        }
